### ISSUE TITLE
GCC 12.1 missing symbol when loading plugin

### ISSUE THREAD

**Issue by boris-kolpackov** [Homebrew/homebrew-core#106394]
### `brew gist-logs <formula>` link OR `brew config` AND `brew doctor` output

```shell
% brew config

HOMEBREW_VERSION: 3.5.6
ORIGIN: https://github.com/Homebrew/brew
HEAD: 0b030b74e2ac518bd0bc4158f96c449198f9957f
Last commit: 30 hours ago
Core tap ORIGIN: https://github.com/Homebrew/homebrew-core
Core tap HEAD: b18dae855942fc9b75ff9741e395481416390433
Core tap last commit: 23 hours ago
Core tap branch: master
HOMEBREW_PREFIX: /usr/local
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 8
Homebrew Ruby: 2.6.8 => /System/Library/Frameworks/Ruby.framework/Versions/2.6/usr/bin/ruby
CPU: octa-core 64-bit penryn
Clang: 13.1.6 build 1316
Git: 2.32.1 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 7.79.1 => /usr/bin/curl
macOS: 12.5-x86_64
CLT: 13.4.0.0.1.1651278267
Xcode: N/A
```

```
% brew info gcc

gcc: stable 11.3.0 (bottled), HEAD
GNU compiler collection
https://gcc.gnu.org/
Not installed
From: https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/gcc.rb
License: GPL-3.0-or-later with GCC-exception-3.1
==> Dependencies
Required: gmp, isl, libmpc, mpfr, zstd
==> Options
--HEAD
	Install HEAD version
```

### Verification

- [X] I ran `brew update` and am still able to reproduce my issue.
- [X] I have resolved all warnings from `brew doctor` and that did not fix my problem.

### What were you trying to do (and why)?

I am attempting to build and load a GCC plugin (it is part of the ODB compiler project).

### What happened (include all command output)?

Everything builds fine but when I attempt to load the plugin, I get this diagnostics:

```
cc1plus: error: cannot load plugin /private/tmp/build-host/odb-2.5.0-b.22.20220718100540.58b7c3add166/odb/odb.so: dlopen(/private/tmp/build-host/odb-2.5.0-b.22.20220718100540.58b7c3add166/odb/odb.so, 0x000A): symbol not found in flat namespace (__Z10pragma_lexPP9tree_nodePj)
```

This only appear to happen with GCC 12.1. I have a build with GCC 11.2 which builds and loads without any problem. To be exact, the GCC 11.2 was also installed from the bottle but on Mac OS 11, not 12. If it helps, I can try to install the latest GCC 11 on exactly the  same machine to confirm, but I am pretty sure it works (I tried GCC 11.3 on an M1 machine, and it does load without any problems).

Also, another piece of information: the same ODB plugin builds and loads fine on Linux with GCC 12.1.

Let me know if you need me to run any commands or provide any additional information.

### What did you expect to happen?

I expected there to be no missing symbols, just like with the previous version of GCC on Mac OS or with this version of GCC on Linux.

### Step-by-step reproduction instructions (by running `brew` commands)

```shell
# The only brew command involved was installing GCC 12.1:

brew install gcc@12
```

**Comment by SMillerDev** [comment]
So what are the brew commands that lead to this error?

**Comment by boris-kolpackov** [comment]
Ok, let me provide some background: GCC has support for compiler plugins. If enabled (which is the case on most platforms, including Homebrew), GCC installs a bunch of headers that are to be included by plugin source code in order to get access to GCC compiler APIs. On all platforms, the "protocol" (as documented somewhere by GCC) is to call GCC with the `-print-file-name=plugin` option and the headers can be found in the `include/` subdirectory of the printed directory. Once the plugin is built (as a shared library), it is loaded by GCC (specifically, `cc1plus` in my case) during compilation and any API symbols that the plugin needs are satisfied by the executable (it is compiled with `-rdynamic` to enable this).

So what seems to happen here is that in Homebrew GCC 12.1 some symbols are missing for some reason in this GCC executable (`cc1plus` in this case). I cannot say for sure whether this is Homebrew issue or upstream issue but the fact that the same plugin builds and loads successfully on Linux with GCC 12.1 would indicate that this could be Mac-specific.

**Comment by carlocab** [comment]
I suggest:

1. creating a minimal reproducer that does not involve ODB at all; and,
2. determining whether this affects only Homebrew `gcc@12` or all macOS builds of GCC 12.1.

This is admittedly quite inconvenient to do, but I don't think this issue has a hope of getting resolved (other than by chance) without both.

**Comment by boris-kolpackov** [comment]
Ok, thanks, I will try to investigate this further.

What builds of GCC for Mac OS are there other than Homebrew?

**Comment by boris-kolpackov** [comment]
This appears to be a x86_64-specific issue, everything loads fine on M1. Is there by any chance something special going on for x86_64? Maybe it's a fat binary or some such?

**Comment by carlocab** [comment]
There is currently no way for anyone else to reproduce this issue, and no one else has commented to say that they are facing a similar problem.

I'm closing this for now. We can reopen this if someone is able to find a simple reproduction of this problem.

**Comment by boris-kolpackov** [comment]
FWIW, we are still investigating this issue.

**Comment by carlocab** [comment]
Sure. We can reopen this when you (or anyone else) has an actionable update on this issue.

**Comment by boris-kolpackov** [comment]
Ok, we've finally tracked this down.

TL;DR: It appears the Mac OS `strip` utility "over-strips" dynamic executables on x86_64 but not M1. This can be fixed by explicitly passing the `-u -r` options to `strip` (which, according to the man page, should be the default for dynamic executables, see below for details). The below patch to the `gcc.rb` formulae fixes the issue (note: the same patch should be applied to `gcc@11.rb`). I am also going to try to report this to Apple and will update this issue if/when I hear back.

```diff
--- gcc.rb.orig 2022-08-26 12:23:33.000000000 +0200
+++ gcc.rb   2022-08-26 12:26:20.456374832 +0200
@@ -109,6 +109,10 @@
     end
 
     mkdir "build" do
+      if OS.mac?
+        ENV["STRIP"] = "strip -u -r"
+      end
+
       system "../configure", *args
       system "make"
 
```

Ok, now the long story:

Firstly, we've tracked this regression to the change from GCC `11.2` to `11.3` where the formulae started doing `install-strip` instead of `install`.

To understand why this is happening I've created a small reproducer that mimics the GCC plugin setup:

```c
/* file: plug.c */

extern int foobar ();

int plugin ()
{
  return foobar ();
}
```

```c
/* file: prog.c */

#include <dlfcn.h>
#include <stdio.h>

int foobar ()
{
  return 0;
}

int main()
{
  void* h = dlopen ("libplug.so", RTLD_LAZY);
  if (h == NULL)
  {
    fprintf(stderr, "could not open plugin library: %s\n", dlerror());
    return 1;
  }

  int (*f) () = dlsym (h, "plugin");
  if (f == NULL)
  {
    fprintf(stderr, "Could not find plugin function: %s\n", dlerror());
    return 1;
  }

  return f ();
}
``` 

The build instructions are as follows:

```
gcc -shared -undefined dynamic_lookup -o libplug.so plug.c
gcc -rdynamic -o prog prog.c -ldl

./prog  # Ok
```

The unstripped executable works fine on both x86_64 and M1. It also works fine everywhere if we strip it with `-u -r`:

```
strip -u -r prog

./prog  # Ok
```

If, however, we strip it without any options, then things start diverging: everything still works fine on M1 but not on x86_64. On M1:

```
strip prog

./prog  # Ok
```

On x86_64:

```
strip prog

./prog

dyld: lazy symbol binding failed: Symbol not found: _foobar
  Referenced from: libplug.so
  Expected in: flat namespace

dyld: Symbol not found: _foobar
  Referenced from: libplug.so
  Expected in: flat namespace

Abort trap: 6
```

The `strip(1)` man page contains the following paragraph which suggest that stripping of this executable without any options should be equivalent to `-u -r`:

>When `strip` is used with no options on an executable file, it checks that file to see if it uses the dynamic link editor.  If it does, the effect of  the  `strip` command  is  the  same as using the `-u` and `-r` options.

We've tried to examine the stripped executable with `nm` on both M1 and x86_64 but couldn't see any difference that would explain this behavior.

We've tested this with `strip` from both CLT 12.4 on Mac OS 10.15.7 and CLT 13.4 on Mac OS 12.5.

**Comment by carlocab** [comment]
CC @fxcoudert

**Comment by fxcoudert** [comment]
@boris-kolpackov thanks for the detailed reproducer and explanation. I wonder if this is a regression in Apple's `strip`, because we never saw this issue before. It would be good to test with a CLT earlier than 12 (but I don't have that at hand).

If you look at the source for `strip`, which is open source: https://opensource.apple.com/source/cctools/cctools-973.0.1/misc/strip.c.auto.html you can see several branches depend on the value of `CPU_TYPE_`, or on `rflag`. One of those might be the issue…

**Comment by iains** [comment]
does your small reproducer exhibit the same issue when clang is used to build it?

**Comment by fxcoudert** [comment]
The issue in the gcc formula is due to the use of `make install-strip`, so let's use `make install` instead and avoid the issue. PR opened at https://github.com/Homebrew/homebrew-core/pull/109063

**Comment by carlocab** [comment]
> does your small reproducer exhibit the same issue when clang is used to build it?

I can reproduce the issue with both `gcc-12` and `clang`.

```
❯ clang -shared -undefined dynamic_lookup -o libplug.so plug.c
❯ clang -rdynamic -o prog prog.c -ldl
❯ ./prog && echo ok
ok
❯ strip -u -r prog
❯ ./prog && echo ok
ok
❯ strip prog
❯ ./prog && echo ok
could not open plugin library: dlopen(libplug.so, 0x0001): symbol not found in flat namespace (_foobar)
```

**Comment by carlocab** [comment]
https://github.com/Homebrew/homebrew-core/pull/109063 has been merged. Do
```
brew update && brew reinstall gcc
```
to get a fixed version.

**Comment by boris-kolpackov** [comment]
@fxcoudert 

> When you report this to Apple, could you please post the FB number here? I will try to follow that with them.

I am not sure what is an "FB number" or how to get one. AFAIK, Apple doesn't have a public-facing issue tracker for things like this but maybe that has changed? What I normally do is email the issue to someone I know who works on Apple Clang, ask them to forward it to the relevant people, and hope for the best. I did that in this case but haven't heard any acknowledgement so it's probably fair to assume things have fallen through the crack. So if you would like to submit it via your channels, please feels free. Here is the above-mentioned email for your reference:

```
We ran into what seems to be a bug in the strip utility that comes
with the Apple Command Line Tools. Last time I hit a bug in Apple
tooling I reported it to Bruno but he is no loner with Apple. If
I need to forward this to someone else, please let me know.

So it appears strip without any options over-strips dynamic
executables on x86_64 but not on M1. Here is a reproducer:


```
/* file: plug.c */

extern int foobar ();

int plugin ()
{
  return foobar ();
}
```


```
/* file: prog.c */

#include <dlfcn.h>
#include <stdio.h>

int foobar ()
{
  return 0;
}

int main()
{
  void* h = dlopen ("libplug.so", RTLD_LAZY);
  if (h == NULL)
  {
    fprintf(stderr, "could not open plugin library: %s\n", dlerror());
    return 1;
  }

  int (*f) () = dlsym (h, "plugin");
  if (f == NULL)
  {
    fprintf(stderr, "Could not find plugin function: %s\n", dlerror());
    return 1;
  }

  return f ();
}
```


And we build and run it like this:


gcc -shared -undefined dynamic_lookup -o libplug.so plug.c
gcc -rdynamic -o prog prog.c -ldl
./prog  # Ok


The unstripped executable works fine on both x86_64 and M1. It also
works fine everywhere if we strip it with -u -r:


strip -u -r prog
./prog  # Ok


If, however, we strip it without any options, then things start diverging:
everything still works fine on M1 but not on x86_64. On M1:


strip prog
./prog  # Ok


On x86_64:


strip prog
./prog

dyld: lazy symbol binding failed: Symbol not found: _foobar
  Referenced from: libplug.so
  Expected in: flat namespace

dyld: Symbol not found: _foobar
  Referenced from: libplug.so
  Expected in: flat namespace

Abort trap: 6


The strip(1) man page contains the following paragraph which suggest that
stripping of this executable without any options should be equivalent to
-u -r:

> When strip is used with no options on an executable file, it checks that
> file to see if it uses the dynamic link editor. If it does, the effect of
> the strip command is the same as using the -u and -r options.

We've tried to examine the stripped executable with nm on both M1 and
x86_64 but couldn't see any difference that would explain this behavior.

We've tested this with strip from both CLT 12.4 on Mac OS 10.15.7 and
CLT 13.4 on Mac OS 12.5.
```

**Comment by carlocab** [comment]
You get an FB number when you file an issue using Apple's [Feedback Assistant](https://feedbackassistant.apple.com).

I suggest reporting this issue there too to get an FB number that we can refer to contacts at Apple so they can follow it up. It's a bit harder to get them to follow up on an email they may or may not be able to track down.

**Comment by boris-kolpackov** [comment]
Thanks for tip! I've submitted the issue and the FB number is `11453932`.

**Comment by carlocab** [comment]
Thanks! You may wish to give your contact at Apple that FB number so they can follow it up internally too.

**Comment by fxcoudert** [comment]
@boris-kolpackov does the fix we merged solve the issue for you?

**Comment by boris-kolpackov** [comment]
@fxcoudert Sorry, I haven't had a chance to test this until now. But after upgrading to 12.2.0 I've hit issue #110673 which is preventing me from confirming this fix. In the meantime I will try with GCC 11.3.0.

**Comment by boris-kolpackov** [comment]
@fxcoudert I've just retested GCC 11.3.0 from the bottle on Mac OS 12.5 and the issue is still there. Perhaps the fix was only applied to `gcc.rb` and not to `gcc@11.rb`? (For GCC 11.3 that formula also switched to using `install-stripped`).

**Comment by fxcoudert** [comment]
@boris-kolpackov fix for gcc@11 in https://github.com/Homebrew/homebrew-core/pull/110706

**Comment by carlocab** [comment]
Incidentally, `strip` breaking binaries on macOS seems to not be that new, as it's also seen here: #100714
